util/git/client.go

```diff
@@ -1142,27 +1143,75 @@ func (m *nativeGitClient) GetCommitNote(sha string, namespace string) (string, e
 	if namespace == "" {
 		namespace = "commit"
 	}
 	ctx := context.Background()
 	ref := "--ref=" + namespace
-	_, err := m.runCmd(ctx, "notes", ref, "add", "-f", "-m", note, sha)
-	if err != nil {
-		return fmt.Errorf("failed to push: %w", err)
-	}
-	if m.OnPush != nil {
-		done := m.OnPush(m.repoURL)
-		defer done()
-	}
+	notesRef := "refs/notes/" + namespace
 
-	err = m.runCredentialedCmd(ctx, "push", "-f", "origin", "refs/notes/"+namespace)
-	if err != nil {
-		return fmt.Errorf("failed to push: %w", err)
+	// Retry up to 5 times to handle concurrent note updates
+	maxRetries := 5
+	var lastErr error
+
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		if attempt > 0 {
+			// Exponential backoff with aggressive jitter to avoid thundering herd
+			baseBackoff := 50 * attempt * attempt // 50ms, 200ms, 450ms, 800ms in milliseconds
+			jitter := rand.Intn(baseBackoff + 1)  // Up to 100% jitter for maximum spread
+			backoff := time.Duration(baseBackoff+jitter) * time.Millisecond
+			log.Debugf("Retrying AddAndPushNote for commit %s (attempt %d/%d) after %v", sha, attempt+1, maxRetries, backoff)
+			time.Sleep(backoff)
+		}
+
+		// Fetch the latest notes BEFORE adding to merge concurrent updates
+		// Use + prefix to force update local ref (safe because we want latest remote notes)
+		_, fetchErr := m.runCmd(ctx, "fetch", "origin", fmt.Sprintf("+%s:%s", notesRef, notesRef))
+		// Ignore "couldn't find remote ref" errors (notes don't exist yet - first time)
+		if fetchErr != nil && !strings.Contains(fetchErr.Error(), "couldn't find remote ref") {
+			log.Debugf("Failed to fetch notes (will continue): %v", fetchErr)
+		}
+
+		// Add note locally (use -f to overwrite if this specific commit already has a note locally)
+		_, err := m.runCmd(ctx, "notes", ref, "add", "-f", "-m", note, sha)
+		if err != nil {
+			return fmt.Errorf("failed to add note: %w", err)
+		}
+
+		if m.OnPush != nil {
+			done := m.OnPush(m.repoURL)
+			defer done()
+		}
+
+		// Push WITHOUT -f flag to avoid overwriting other notes
+		err = m.runCredentialedCmd(ctx, "push", "origin", notesRef)
+		if err == nil {
+			if attempt > 0 {
+				log.Debugf("AddAndPushNote succeeded after %d retries for commit %s", attempt, sha)
+			}
+			return nil
+		}
+
+		lastErr = err
+		log.Debugf("AddAndPushNote push failed (attempt %d/%d): %v", attempt+1, maxRetries, err)
+
+		// Check if this is a retryable error
+		errStr := err.Error()
+		isRetryable := strings.Contains(errStr, "fetch first") || // Remote updated after our fetch (concurrent push completed between our fetch and push)
+			strings.Contains(errStr, "reference already exists") || // Concurrent push is holding the lock (git server-side lock)
+			strings.Contains(errStr, "incorrect old value") || // Git detected our local ref is stale (concurrent update)
+			strings.Contains(errStr, "failed to update ref") // Generic ref update failure that may include transient issues
+
+		if !isRetryable {
```

Comment:
I think we can improve the code a bit here 
we can introduce a slice with all the `retryable` error strings and then just use `slices.Contains(things, "foo")` intead of this `ugly` if block 



```suggestion
	retryableErrors := []string{
		"fetch first",              // Remote updated after our fetch (concurrent push completed between our fetch and push)
		"reference already exists", // Concurrent push is holding the lock (git server-side lock)
		"incorrect old value",      // Git detected our local ref is stale (concurrent update)
		"failed to update ref",     // Generic ref update failure that may include transient issues
	}

	if slices.Contains(retryableErrors, errStr)
```